### ISSUE TITLE
function schemas that can use input values in output schemas

### ISSUE THREAD

**Issue by stevebuik** [metosin/malli#936]
```clojure
(mx/defn tall-people :- [:vector [:map [:height :int]]]
  [{:keys [min-height]} :- [:map [:min-height :int]]]
  [{:height 200}
   {:height 150}])

(tall-people {:min-height 180})
```
=> should throw exception when instrumented

spec1 s/fdef can do this so parity or better would be ideal. what are strengths/weaknesses of s/fdef ?

**Comment by stevebuik** [comment]
thanks to chatgpt, here's the fdef equivalent for comparison

```clojure
(s/def ::height int?)
(s/def ::min-height int?)
(s/def ::person (s/keys :req-un [::height]))
(s/def ::tall-people-args (s/keys :req-un [::min-height]))
(s/def ::tall-people-result (s/coll-of ::person :kind vector?))

(defn tall-people [args]
  (let [{:keys [min-height]} args]
    [{:height 200}
     {:height 150}]))

(s/fdef tall-people
        :args (s/cat :args ::tall-people-args)
        :ret (s/coll-of ::person :kind vector?)
        :fn (fn [{:keys [args ret]}]
              (let [{:keys [min-height]} (:args args)]
                (every? #(>= (:height %) min-height) ret))))

(stest/instrument)

(tall-people {:min-height 180})
```
although return values are not checked even when instrumented I believe

I hope Malli can be more terse. what other improvements can we make on this?

**Comment by ikitommi** [comment]
this can be made to work:

```clojure
(def Input [:map [:min-height :int]])

(defn guard [{:keys [input output]}]
  (every? #(>= (:height %) (:min-height input)) output))

(mx/defn kikka :- [:vector Input]
  {:fn guard}
  [_ :- Input]
  ;; body
  [{:height 200}
   {:height 150}])
```

**Comment by stevebuik** [comment]
Looks good. Can the Input and other schemas be provided by the usual
options with a registry with this design?

We have moved to using registries everywhere

**Comment by knubie** [comment]
I think this is related to / dup of #764

**Comment by allentiak** [comment]
It would be nice if something like this could work.
For this, named params are awesome.

```clojure
(defn area-of-a-square
[len]
{:malli/schema [:=> [:catn [:len nat-int?]]
     [:catn [:area
              [:and
               nat-int?
               [:fn (fn [{:keys [area len]}]
                  (>= area len))]]]]]}
(* len len))
```
With clojure.spec, I would write it in this way:

```clojure
(defn area-of-square
  [len]
  (* len len))

(s/fdef area-of-square
  :args (s/cat :len nat-int?)
  :ret nat-int?
  :fn (fn [{:keys [ret len]}]
        (>= ret len))]]]]]}


```

**Comment by ikitommi** [comment]
quick poke on this.

<img width="610" alt="Screenshot 2024-01-24 at 19 00 01" src="https://github.com/metosin/malli/assets/567532/c0761aaf-6d99-4caa-969c-0e730be5b74c">

**Comment by ikitommi** [comment]
fixed in #1000.
